File: Data Structure/Union find/LET_no1697_Checking Existence of Edge Length Limited Paths.py
class Solution:
    def distanceLimitedPathsExist(self, n: int, edgeList: List[List[int]], queries: List[List[int]]) -> List[bool]:
        res = [None] * len(queries)
        edgeList.sort(key=lambda x: x[2])
        queries = sorted([q + [i] for i, q in enumerate(queries)], key=lambda x: x[2])
        
        un = UnionFind(n)
        i = 0
        
        for a, b, limit, idx in queries:
            while i < len(edgeList) and edgeList[i][2] < limit:
                x, y, d = edgeList[i]
                un.union(x, y)
                i += 1
            res[idx] = un.find(a) == un.find(b)
        
        return res
    
    
class UnionFind:
    def __init__(self, N):
        self.root = list(range(N))
    
    def find(self, x):
        if self.root[x] != x:
            self.root[x] = self.find(self.root[x])
        return self.root[x]
    
    def union(self, a, b):
        rx, ry = self.find(a), self.find(b)
        if rx != ry:
            self.root[ry] = rx
            return
    

File: Data Structure/Union find/Hard/LET_no1697_Checking Existence of Edge Length Limited Paths.cpp
class UnionFind {
private:
    vector<int> parent, rank;
public:
    UnionFind(int size) {
        parent.resize(size, 0);
        for (int i = 0; i < size; i++)
            parent[i] = i;
    }

    int find(int x) {
        if (parent[x] != x)
            parent[x] = find(parent[x]);

        return parent[x];
    }

    void join(int x, int y) {
        int px = find(x), py = find(y);
        if (px != py)
            parent[px] = py;
    }
};

class Solution {
public:
    vector<bool> distanceLimitedPathsExist(int n, vector<vector<int>>& edgeList, vector<vector<int>>& queries) {
        UnionFind uf(n);

        for (int i = 0; i < queries.size(); i++)
            queries[i].push_back(i);

        sort(queries.begin(), queries.end(), [](auto& l, auto& r) { return l[2] < r[2]; } );
        sort(edgeList.begin(), edgeList.end(), [](auto& l, auto& r) { return l.back() < r.back(); } );

        int i = 0;
        vector<bool> res(queries.size(), false);
        for (vector<int>& q : queries) {
            while (i < edgeList.size() && edgeList[i][2] < q[2]) {
                uf.join(edgeList[i][0], edgeList[i][1]);
                i++;
            }

            res[q.back()] = uf.find(q[0]) == uf.find(q[1]);
        }

        return res;
    }
};